### PR TITLE
[bookmarks] [ios] Fix bookmarks categories sorting bug

### PR DESCRIPTION
**Pull request by kirylkaveryn**
Closes https://github.com/organicmaps/organicmaps/issues/3475

This issue contains two bugs:
## 1. Invalid sorting
1. To display the categories list in the ios we use MWMBookmarksManager's :
` (NSArray<MWMBookmarkGroup *> *)userCategories;`
  which calls c++ BoormarkManager's
`auto const & list = self.bm.GetBmGroupsIdList(); `
  which returns
`kml::GroupIdCollection const & GetBmGroupsIdList() const { return m_bmGroupsIdList; }`

`m_bmGroupsIdList` is a sorted list of categories ids
and sorting happens in the method:
```cpp
void BookmarkManager::UpdateBmGroupIdList()
{
  CHECK_THREAD_CHECKER(m_threadChecker, ());
  size_t const count = m_categories.size();

  // All this routine to get sorted list by last modified time.

  using PairT = std::pair<kml::MarkGroupId, BookmarkCategory const *>;
  std::vector<PairT> vec;
  vec.reserve(count);
  for (auto const & [markGroupId, categoryPtr] : m_categories)
    vec.emplace_back(markGroupId, categoryPtr.get());

  std::sort(vec.begin(), vec.end(), [](PairT const & lhs, PairT const & rhs)
  {
    return lhs.second->GetLastModifiedTime() > rhs.second->GetLastModifiedTime();
  });

  m_bmGroupsIdList.clear();
  m_bmGroupsIdList.resize(count);
  for (size_t i = 0; i < count; ++i)
    m_bmGroupsIdList[i] = vec[i].first;
}

```

This method is called only in:
- `CreateBookmarkCategory`
- `DeleteBmCategory`

So when the user adds/removes/updates bookmarks `m_bmGroupsIdList` becomes invalid (in case of sorting).

How to reproduce:
- Add the bookmark to the group
- Open the Bookmarks screen
- The group will be at the same place as earlier
- Terminate app
- Open app
- Group will be at the top of the list (`CreateBookmarkCategory` was called on init so `m_bmGroupsIdList` will be correct until the next editing)

Simulator Screen Recording - iPhone 15 Pro - 2024-03-07 at 13 59 04

Solution:
The new method is implemented always to retrieve valid sorted list of IDs:
 `std::vector<kml::MarkGroupId> BookmarkManager::GetSortedBmGroupIdList() const` in the core
and 
`- (NSArray<MWMBookmarkGroup *> *)sortedUserCategories` in the ios

### CODE REVIEW

File: kml/serdes.cpp [hunk 1]
```diff
@@ -306,11 +306,13 @@ void SaveCategoryExtendedData(KmlWriter::WriterWrapper & writer, CategoryData co
...
-  {
-    writer << indent << "<mwm:lastModified>" << TimestampToString(categoryData.m_lastModified)
-           << "</mwm:lastModified>\n";
+  Timestamp lastModifiedTime;
+  if (categoryData.m_lastModified != Timestamp()) {
+    lastModifiedTime = categoryData.m_lastModified;
+  } else {
+    lastModifiedTime = kml::TimestampClock::now();
   }
+  writer << indent << "<mwm:lastModified>" << TimestampToString(lastModifiedTime) << "</mwm:lastModified>\n";
```
Comment (vng) [inline kml/serdes.cpp:315]: I suspect that this change breaks all test's constrains. We do Save X, but Read Y.
We change the inner state in the Save function which is not good.

I still haven't read all the PR descriptions, but if this is the real error case, we _should assign_ a valid timestamp somewhere else: e.g. in the Import, Create, Modify functions, but not in Save.

**Review by vng — CHANGES_REQUESTED** [review]
Please, keep our code style in C++
Do not put braces {} for one line expressions. Also we use new-line braces, not KR-style.

File: map/bookmark_manager.hpp [hunk 2]
```diff
@@ -277,6 +277,7 @@ class BookmarkManager final
   kml::MarkGroupId GetCategoryId(std::string const & name) const;
 
   kml::GroupIdCollection const & GetBmGroupsIdList() const { return m_bmGroupsIdList; }
```
Comment (biodranik) [inline map/bookmark_manager.hpp:279]: There are cases where an unsorted list is ok, right? What are these cases?
```suggestion
  kml::GroupIdCollection const & GetUnsortedBmGroupsIdList() const { return m_bmGroupsIdList; }
```

File: map/bookmark_manager.hpp [hunk 3]
```diff
@@ -277,6 +277,7 @@ class BookmarkManager final
...
+  kml::GroupIdCollection GetSortedBmGroupIdList() const;
```
Comment (biodranik) [inline map/bookmark_manager.hpp:280]: How often is this method called? Does it make sense to sort and cache sorted lists in m_bmGroupsIdList, to avoid resorting it again and again each time for displaying?

Does it make sense to insert sorting into each bookmark/track modification operation (e.g. sort when any category has updated its last modified date)? Then there is no need to sort list of categories each time when retrieving it. By the way, the sorting by the last modified field in such cases for the vector can be implemented without a full sorting: scan vector to find the index of the modified category, and then move it to the beginning of the vector.

File: map/bookmark_manager.cpp [hunk 4]
```diff
@@ -2207,7 +2207,25 @@ void BookmarkManager::UpdateBmGroupIdList()
...
+  using PairT = std::pair<kml::MarkGroupId, BookmarkCategory const *>;
+  std::vector<PairT> vec;
+  vec.reserve(count);
+  for (auto const & [markGroupId, categoryPtr] : m_categories)
+    vec.emplace_back(markGroupId, categoryPtr.get());
+
+  m_bmGroupsIdList.clear();
+  m_bmGroupsIdList.resize(count);
+  for (size_t i = 0; i < count; ++i)
+    m_bmGroupsIdList[i] = vec[i].first;
```
Comment (biodranik) [inline map/bookmark_manager.cpp:2219]: Will m_bmGroupsIdList always contain numbers from 1 to N, where N is a total categories count?

**Review by biodranik — COMMENTED** [review]
Note the failed CI tests.

There is a fundamental issue that categories are stored internally in the std::map for faster access by id, and there is always a sorted copy of ids in that map that is queried by clients to display categories and for other tasks.

Let's list all cases when a list of categories is accessed by the client code. I saw these two:
1. Displaying a sorted list of categories in the Bookmarks and Tracks dialog
2. Exporting all bookmarks (the sort order is not important here).

What else?

If sorting is always called for each UI query, then there may not be a need to keep a vector with ids at all. A client may query all categories explicitly either without sorting or with some specific sort order. 

The approach above has the disadvantage of many unnecessary sorting calls (depending on how frequently the list is queried by clients). If it's an issue (think about tens/hundreds of lists), then it may be more efficient to sort groups when any last modified field has changed, and return to the UI the already sorted vector.

That is why it is important to understand how these lists are used now on both iOS and Android.

File: kml/serdes.cpp (hunk 1 above)
Comment (cyber-toad) [reply kml/serdes.cpp:315]: > @cyber-toad Am I right that m_lastModified for lists (categories) is not initialized now neither for KML nor for GPX imports? If yes, can you please help with a fix?

@biodranik yes, it is not used in parser an reset to current time is done later in UpdateLastModifiedTime. So we can have it parsed there if it is in file, but if we will initialize it to default (current) time in parser we will have different result for two executions of parser on the same source, that will make it more difficult to test (we will need to pass current time for test from test code). Or there is a better place than parser for defaulting?

Comment (biodranik) [reply kml/serdes.cpp:315]: Does it make sense to add `Platform::GetFileModificationTime` in the same way as `Platform::GetFileCreationTime`, and use it during the import? Or maybe even rename it to `GetFileTime` and pass an enum `ECreationTime` and `EModificationTime` as a parameter?

Comment (cyber-toad) [reply kml/serdes.cpp:315]: I'm not sure that these values are stable, for example:

```
~/dev/organicmaps/data/gpx_test_data$ stat route.gpx
  File: route.gpx
...
Access: 2024-03-08 10:00:05.705673721 +0100
Modify: 2024-01-14 15:45:23.712451946 +0100
Change: 2024-01-14 15:45:23.712451946 +0100
 Birth: 2024-01-14 15:45:23.712451946 +0100
```

So birth/change/modify were set during git pull, so will be changed if another dev made a checkout and tests fail?

Comment (biodranik) [reply kml/serdes.cpp:315]: The test can import a file without that field, and compare that field to manually retrieved file modification date.

Comment (biodranik) [reply kml/serdes.cpp:315]: Comparison with a modification time stored in the file should be easy - it should not change during the import. Then we can preserve the time for backups.

Comment (kirylkaveryn) [reply kml/serdes.cpp:315]: @cyber-toad can you please help me to fix this importing issue?

**Comment by kirylkaveryn** [comment]
> Note the failed CI tests.
> 
> There is a fundamental issue that categories are stored internally in the std::map for faster access by id, and there is always a sorted copy of ids in that map that is queried by clients to display categories and for other tasks.
> 
> Let's list all cases when a list of categories is accessed by the client code. I saw these two:
> 
> 1. Displaying a sorted list of categories in the Bookmarks and Tracks dialog
> 2. Exporting all bookmarks (the sort order is not important here).
> 
> What else?
> 
> If sorting is always called for each UI query, then there may not be a need to keep a vector with ids at all. A client may query all categories explicitly either without sorting or with some specific sort order.
> 
> The approach above has the disadvantage of many unnecessary sorting calls (depending on how frequently the list is queried by clients). If it's an issue (think about tens/hundreds of lists), then it may be more efficient to sort groups when any last modified field has changed, and return to the UI the already sorted vector.
> 
> That is why it is important to understand how these lists are used now on both iOS and Android.

UI uses this property in iOS:
- in the Bookmarks list
- in the Group changing screen

And the frequency of use depends on how the user interacts with the app...

I think that the sorting operation isn't too expensive when we are talking about 1<100 categories. It's not thousands.

And we should choose:
1) sort the IDs array on _every_ bookmark/track add/change/delete action
pros:
- always ready-to-use vector
cons:
- the name of the property is not self-describing (or we should use two variables)
- interfere implementation of sorting filters (as date/name/type etc) feature
- sorting should be done even if we don't need it on EVERY bookmark/track change
OR 
2) sort only when necessary using some sorting options when we show tableView with bookmarks
pros:
- sorted vector is created only when we need it
- separated usage: non-sorted and sorted arrays
- filters may be implemented
cons:
- may impact performance in some extremely rare situations
 
What do users make more often: open the 10 times bookmarks screen or add the 10 bookmarks?
In both cases, sorting happens 10 times, but in the 1st (current) implementation pre-sorting is more like a side effect rather than expected behavior.

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.h [hunk 5]
```diff
@@ -92,6 +92,7 @@ NS_SWIFT_NAME(BookmarksManager)
 - (BOOL)areNotificationsEnabled;
 
 - (NSArray<MWMBookmarkGroup *> *)userCategories;
+- (NSArray<MWMBookmarkGroup *> *)sortedUserCategories;
```
[the remark this answers is not in the document]
Comment (kirylkaveryn) [reply iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.h:95]: I think that should be at least `default (by date)` and `by name` options as implemented in the bookmarks.

File: map/bookmark_manager.hpp (hunk 2 above)
Comment (kirylkaveryn) [reply map/bookmark_manager.hpp:279]: This case is ok when we want to check the count of groups. For example in the BookmarksListInteractor:
```swift 
  func canDeleteGroup() -> Bool {
    bookmarksManager.userCategories().count > 1
  }
```

This property is accessed a few times: 
when the sorted group list should be displayed 
and to check is the array is empty:
image

File: map/bookmark_manager.cpp (hunk 4 above)
Comment (kirylkaveryn) [reply map/bookmark_manager.cpp:2219]: Yes it will. This property is updated on the `addCategort` and `deleteCategory` actions (in the current implementation).
Or maybe I don't understand your question...

**Comment by biodranik** [comment]
There is also option 3:
- Store a "needsSorting" flag when any bookmark is modified
- Sort lists only once if this flag is set when lists are retrieved by UI functions that _must_ display sorted lists.

Users also want to have sorting by name implemented for lists. What would be the best approach in this case?

File: map/bookmark_manager.hpp (hunk 2 above)
Comment (biodranik) [reply map/bookmark_manager.hpp:279]: This is not reliable compared to `bookmarksManager.userCategoriesCount()`

**Comment by kirylkaveryn** [comment]
> Users also want to have sorting by name implemented for lists. What would be the best approach in this case?
@biodranik 

We can use the same approach as already used for bookmarks (but simplified):
```swift
- (void)sortBookmarks:(MWMMarkGroupID)groupId
          sortingType:(MWMBookmarksSortingType)sortingType
             location:(CLLocation *)location
           completion:(SortBookmarksCompletionBlock)completion {
  self.bm.SetLastSortingType(groupId, convertSortingTypeToCore(sortingType));
  m2::PointD myPosition = m2::PointD::Zero();

  if (sortingType == MWMBookmarksSortingTypeByDistance) {
    if (!location) {
      completion(nil);
      return;
    }
    myPosition = mercator::FromLatLon(location.coordinate.latitude, location.coordinate.longitude);
  }

  auto const sortId = ++self.lastSortId;
  __weak auto weakSelf = self;

  BookmarkManager::SortParams sortParams;
  sortParams.m_groupId = groupId;
  sortParams.m_sortingType = convertSortingTypeToCore(sortingType);
  sortParams.m_hasMyPosition = location != nil;
  sortParams.m_myPosition = myPosition;
  sortParams.m_onResults = [weakSelf, sortId, completion] (BookmarkManager::SortedBlocksCollection &&sortedBlocks,
                                                           BookmarkManager::SortParams::Status status) {
    __strong auto self = weakSelf;
    if (!self || sortId != self.lastSortId)
      return;

    switch (status) {
      case BookmarkManager::SortParams::Status::Completed: {
        NSMutableArray *result = [NSMutableArray array];
        for (auto const &sortedBlock : sortedBlocks) {
          NSMutableArray *bookmarks = nil;
          if (sortedBlock.m_markIds.size() > 0) {
            bookmarks = [NSMutableArray array];
            for (auto const &markId : sortedBlock.m_markIds) {
              [bookmarks addObject:[[MWMBookmark alloc] initWithMarkId:markId
                                                          bookmarkData:self.bm.GetBookmark(markId)]];
            }
          }
          NSMutableArray *tracks = nil;
          if (sortedBlock.m_trackIds.size() > 0) {
            tracks = [NSMutableArray array];
            for (auto const &trackId : sortedBlock.m_trackIds) {
              [tracks addObject:[[MWMTrack alloc] initWithTrackId:trackId trackData:self.bm.GetTrack(trackId)]];
            }
          }
          [result addObject:[[MWMBookmarksSection alloc] initWithTitle:@(sortedBlock.m_blockName.c_str())
                                                             bookmarks:bookmarks
                                                                tracks:tracks]];
        }
        completion([result copy]);
        break;
      }
      case BookmarkManager::SortParams::Status::Cancelled:
        completion(nil);
        break;
    }
  };

  self.bm.GetSortedCategory(sortParams);
}

typedef NS_ENUM(NSInteger, MWMBookmarksSortingType) {
  MWMBookmarksSortingTypeByType,
  MWMBookmarksSortingTypeByDistance,
  MWMBookmarksSortingTypeByTime
} NS_SWIFT_NAME(BookmarksSortingType);

```

**Comment by biodranik** [comment]
A similar approach can be used, right. The main question: would it be better to do it in C++ to reuse also on Android and reuse/test in the desktop build of OM (btw, did you run it?)?

Looks like the sorting of bookmarks is implemented in the C++ core: https://github.com/organicmaps/organicmaps/pull/7182

Can the same code be reused for sorting _bookmarks_ in iOS? Why iOS sorts in ObjC instead of using C++?

**Comment by kirylkaveryn** [comment]
> A similar approach can be used, right. The main question: would it be better to do it in C++ to reuse also on Android and reuse/test in the desktop build of OM (btw, did you run it?)?
> 
> Looks like the sorting of bookmarks is implemented in the C++ core: #7182
> 
> Can the same code be reused for sorting _bookmarks_ in iOS? Why iOS sorts in ObjC instead of using C++?



> A similar approach can be used, right. The main question: would it be better to do it in C++ to reuse also on Android and reuse/test in the desktop build of OM (btw, did you run it?)?
> 
> Looks like the sorting of bookmarks is implemented in the C++ core: #7182
> 
> Can the same code be reused for sorting _bookmarks_ in iOS? Why iOS sorts in ObjC instead of using C++?

@biodranik this iOS code isn't sorting. Its look huge, but sorting happens in the cpp line:
`self.bm.GetSortedCategory(sortParams);`

All previous code is the cpp sorting `parameters configuration`, that should be passed into the cpp method to enable sorting.

**Comment by biodranik** [comment]
Wow, it looks complex ) Ok, good. So how can we proceed with this PR then?

**Comment by kirylkaveryn** [comment]
> Wow, it looks complex ) Ok, good. So how can we proceed with this PR then?

We can implement sortings options in the next PR.

For now we can: 
1. leave both
```cpp
  kml::GroupIdCollection const & GetBmGroupsIdList() const { return m_bmGroupsIdList; }
  kml::GroupIdCollection GetSortedBmGroupIdList() const;
```
In the second one we will add options as part of the other PR. It wlll be easier to implement options on this base.

2. If @cyber-toad can help me to fix timestamp inititlization bug, it will be cool!

File: iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm [hunk 6]
```diff
@@ -651,6 +641,10 @@ - (MWMBookmarkGroup *)categoryWithId:(MWMMarkGroupID)groupId {
   return [[MWMBookmarkGroup alloc] initWithCategoryId:groupId bookmarksManager:self];
 }
 
+- (uint16_t)userCategoriesCount {
```
Comment (biodranik) [inline iphone/CoreApi/CoreApi/Bookmarks/MWMBookmarksManager.mm:644]: Why uint16_t instead of size_t?

File: android/app/src/main/cpp/app/organicmaps/bookmarks/data/BookmarkManager.cpp [hunk 7]
```diff
@@ -637,11 +637,21 @@ Java_app_organicmaps_bookmarks_data_BookmarkManager_nativeGetBookmarkCategories(
...
   auto const & bm = frm()->GetBookmarkManager();
-  auto const & ids = bm.GetBmGroupsIdList();
+  auto const & ids = bm.GetSortedBmGroupIdList();
 
   return MakeCategories(env, ids);
 }
 
+JNIEXPORT jint JNICALL
+Java_app_organicmaps_bookmarks_data_BookmarkManager_nativeGetBookmarkCategoriesCount(JNIEnv *env,
+                                                                                jobject)
```
Comment (biodranik) [inline android/app/src/main/cpp/app/organicmaps/bookmarks/data/BookmarkManager.cpp:647]:
```suggestion
Java_app_organicmaps_bookmarks_data_BookmarkManager_nativeGetBookmarkCategoriesCount(JNIEnv *env, jobject)
```

File: map/bookmark_manager.cpp [hunk 8]
```diff
@@ -2220,10 +2238,11 @@ void BookmarkManager::UpdateBmGroupIdList()
     return lhs.second->GetLastModifiedTime() > rhs.second->GetLastModifiedTime();
   });
 
-  m_bmGroupsIdList.clear();
-  m_bmGroupsIdList.resize(count);
-  for (size_t i = 0; i < count; ++i)
-    m_bmGroupsIdList[i] = vec[i].first;
+  for (size_t i = 0; i < count; ++i) {
+    sortedList.push_back(vec[i].first);
+  }
```
Comment (biodranik) [inline map/bookmark_manager.cpp:2243]: One-liners without bracers. If with bracers, they're on a new line in C++ (it was the same in ObjC until Swift arrived).
```suggestion
  for (size_t i = 0; i < count; ++i)
    sortedList.push_back(vec[i].first);
```

**Comment by biodranik** [comment]
https://github.com/organicmaps/organicmaps/actions/runs/8233476612/job/22513010400?pr=7533#step:8:708

File: kml/serdes.cpp (hunk 1 above)
Comment (cyber-toad) [reply kml/serdes.cpp]: @biodranik so you suggest to enrich all current test gpx/kml/kmb data with this timestamp to make existing tests stable and add a couple of new tests where this field is empty and test compares defaulted value with date of file extracted in tests?

@kirylkaveryn I can have a look on it, but I'm a bit busy now so can't provide any ETA. But I see that you extracted it to separate task so will have a look once there is a time.

File: map/map_tests/bookmarks_test.cpp [hunk 9]
```diff
@@ -202,9 +202,9 @@ UNIT_CLASS_TEST(Runner, Bookmarks_ImportKML)
                                  LoadKmlData(MemReader(kmlString, strlen(kmlString)), KmlFileType::Text));
   TEST(kmlDataCollection.back().second, ());
   bmManager.CreateCategories(std::move(kmlDataCollection));
-  TEST_EQUAL(bmManager.GetBmGroupsIdList().size(), 1, ());
+  TEST_EQUAL(bmManager.GetUnsortedBmGroupsIdList().size(), 1, ());
```
[the remark this answers is not in the document]
Comment (kirylkaveryn) [reply map/map_tests/bookmarks_test.cpp]: Fixed!

**Review by biodranik — APPROVED** [review]
Ok, let's finish this one. A TODO comment about duplicate sorting call on iOS would be helpful.

File: map/bookmark_manager.cpp [hunk 10]
```diff
@@ -2207,16 +2207,14 @@ void BookmarkManager::UpdateBmGroupIdList()
...
   m_unsortedBmGroupsIdList.clear();
   m_unsortedBmGroupsIdList.resize(count);
-  for (size_t i = 0; i < count; ++i)
-    m_unsortedBmGroupsIdList[i] = vec[i].first;
+  int i = 0;
+  for (auto const & [markGroupId, _] : m_categories)
+  {
+    m_unsortedBmGroupsIdList[i] = markGroupId;
+    i++;
+  }
```
Comment (biodranik) [inline map/bookmark_manager.cpp:2217]: Soon you'll become a guru of C++ )
```suggestion
  size_t i {0};
  for (auto const & [markGroupId, _] : m_categories)
    m_unsortedBmGroupsIdList[i++] = markGroupId;
```

Comment (kirylkaveryn) [reply map/bookmark_manager.cpp]: This is tough!

**Review by biodranik — APPROVED** [review]
Tadam! Finally, we made it :) Thanks!

**Review by vng — APPROVED** [review]
